### PR TITLE
[Text] Multiple text processing fixes

### LINKED ISSUE #15597: Add Font Features, Languages, and Name IDs on GlyphTypeface
### Is your feature request related to a problem? Please describe.

Querying the capabilities and informational data of a font is not directly supported by Avalonia. Besides apps that load fonts at runtime, such support could also open up new avenues for DevTools improvements.

### Describe the solution you'd like

These should be added to `GlyphTypeface`:

* `Feature` is exposed by HarfBuzzSharp and exists as `FontFeature(Collection)` in Avalonia.
* `Language` is exposed by HarfBuzzSharp.
* Name IDs (`OpenTypeNameId`) are not exposed and need to be added to HarfBuzzSharp's `Face` via `hb_ot_name_get_utf8`.

### Describe alternatives you've considered

_No response_

### Additional context

Online tools such as https://fontdrop.info/#/?darkmode=true exposes this information in a usable way. This is sufficient for working with fonts with characteristics known at compile-time and may help others until Avalonia provides better information.

### PR DESCRIPTION
**Pull request by Gillibald**
## What does the pull request do?

It adds basic font table loading capabilities so we can process the font metadata ourselves without relying on Skia.

- Adds name table processing for localized font family names
- Adds OS/2 HorizontalHeadTable for font metrics calculation
- Adds supported font feature list to IGlyphTypeface2
- Fixes baseline calculation
- Fixes TextLineImpl baseline for drawable runs

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Fixed issues

Fixes: #15597

### CODE REVIEW

**Comment by robloo** [comment]
I don't think the SixLabors code can be used in Avalonia:

 1. SixLabors dual licenses. It is NOT for commercial use. Avalonia however is allowed for commercial use. So if anyone uses Avalonia in a commercial setting they can't actually use SixLabors code under Apache 2.
 2. Apache is less permissive than MIT. I'm not a lawyer but I wouldn't mix and match the code per se. I wouldn't copy in Uno Platform code for a control for example. It it was an entire component/layer of Avalonia it might be easier to manage.

We need direct written permission from SixLabors to use their code here I think.

**Comment by Gillibald** [comment]
> I don't think the SixLabors code can be used in Avalonia:
> 
> 1. SixLabors dual licenses. It is NOT for commercial use. Avalonia however is allowed for commercial use. So if anyone uses Avalonia in a commercial setting they can't use SixLabors code under Apache 2.
> 2. Apache is less permissive than MIT. I'm not a lawyer but I wouldn't mix and match the code per se. I wouldn't copy in Uno Platform code for a control for example. It it was an entire component/layer of Avalonia.
> 
> We need direct written permission from SixLabors to use their code here I think.

MIT license is compatible with the Apache 2.0 license and the code was taken from a point in time when no dual license existed. We can ask @JimBobSquarePants for permission if you want to clarify things.

**Comment by robloo** [comment]
@Gillibald If you copied the code from before the license change a few years ago there is no show-stopper license conflict. 

Perhaps a date should be noted in the header of the source code to avoid any potential concerns for posterity. (Ported from ... at commit <hash> dated <date>

While MIT and Apache 2.0 code is usable together the requirements are quite different. I think we need to note in documents that some portions of source code are under a different license. End apps may need to distribute a copy of the Apache 2.0 license as well. Not saying anything of the patent grants and prominant file change notices which Avalonia is now going to have to follow. These complications are why I really, really try to avoid mixing licenses in source code that is side-by-side.

**Comment by JimBobSquarePants** [comment]
> > I don't think the SixLabors code can be used in Avalonia:
> > 
> > 1. SixLabors dual licenses. It is NOT for commercial use. Avalonia however is allowed for commercial use. So if anyone uses Avalonia in a commercial setting they can't use SixLabors code under Apache 2.
> > 2. Apache is less permissive than MIT. I'm not a lawyer but I wouldn't mix and match the code per se. I wouldn't copy in Uno Platform code for a control for example. It it was an entire component/layer of Avalonia.
> > 
> > We need direct written permission from SixLabors to use their code here I think.
> 
> MIT license is compatible with the Apache 2.0 license and the code was taken from a point in time when no dual license existed. We can ask @JimBobSquarePants for permission if you want to clarify things.

Not that you need it but you have my blessing.

**Comment by Gillibald** [comment]
Potentially fixes: #10658 and #15768

**Comment by robloo** [comment]
For my reported issue I use the ComboBox in the ControlCatalog with individual fonts. Stepping through you can visually see if things are aligned or not. I can take a look later too.

Not sure how to unit test this but it probably can be done with text formatter APIs and some carefully chosen font and strings. The font would have to exist on the test PC though so not sure that's a good idea to add at this point.

**Comment by AngryCarrot789** [comment]
I can confirm that `11.2.999-cibuild0048686-alpha` fixes the font rendering for Oxanium.
image

Consolas looks weird (in my app on the right) compared to windows notepad (left), not sure if it's just me though
image

**Comment by Gillibald** [comment]
The left side is somehow stretched vertically

File: src/Avalonia.Base/Media/TextFormatting/TextLineImpl.cs [hunk 1]
```diff
@@ -1288,11 +1288,10 @@ private TextLineMetrics CreateLineMetrics()
...
                             }
 
-                            if (ascent > -drawableTextRun.Baseline)
-                            {
-                                ascent = -drawableTextRun.Baseline;
-                            }
-
+                            var offset = drawableTextRun.Baseline - descent + ascent;
+                          
+                            ascent -= offset;
```
Comment (jp2masa) [inline src/Avalonia.Base/Media/TextFormatting/TextLineImpl.cs:1293]: This is equivalent to setting `ascent = descent - drawableTextRun.Baseline`, I think? Just to confirm, the previous value of `ascent` is supposed to be discarded?

Comment (Gillibald) [reply src/Avalonia.Base/Media/TextFormatting/TextLineImpl.cs:1293]: Offset is just there for debugging. Will rewrite those lines.

**Review by maxkatz6 — APPROVED** [review]

**Comment by robloo** [comment]
I've checked this with the font ComboBox in ControlCatalog:

capturegif

There are some major improvements for sure. It functions better than the alternative implementation here as well: https://github.com/AvaloniaUI/Avalonia/pull/15344#issuecomment-2053713998.

I'm still a little suspicious of some fonts though. Courier New for example should likely vertically center, Consolas as well. I'm not sure this is 100% yet but don't know enough about this area to be more specific.

**Comment by kurobirds** [comment]
Hello @Gillibald, Can you re-check with FontStretch property? It not render correctly compare to version 11.0.10.

Here is my code to reproduce:
```cs
<TextBlock FontWeight="Heavy"
           FontStretch="Normal"
           FontSize="48"
           LineHeight="56"
           Text="Normal" />
<TextBlock FontWeight="Heavy"
           FontStretch="Condensed"
           FontSize="48"
           LineHeight="56"
           Text="Condensed" />
<TextBlock FontWeight="Heavy"
           FontStretch="Expanded"
           FontSize="48"
           LineHeight="56"
           Text="Expanded" />
```

image

**Comment by Gillibald** [comment]
What font are you using?

**Comment by kurobirds** [comment]
I'm using default font face `{compositefont:fonts:Inter#Inter, $Default#Inter, $Default}`, it also happened on UniversNextPro font.

**Comment by Gillibald** [comment]
Screenshot 2024-06-05 125944
`<TextBlock FontWeight="Black" FontStretch="Condensed" FontFamily="/#Scan" Text="Hello World" FontSize="44"/>`
Some condensed heavy font from the internet. Works just fine. I don't think any system comes with a font that covers this combination.

You request a heavy condensed font but it does not find one so the weight is simulated and the stretch is ignored.

**Comment by kurobirds** [comment]
The issue is that in version `11.0.10`, it can detect my **FontStretch**, but in version `11.2.999-cibuild0048720-alpha`, it cannot. Can you check this with the repository: [AvaloniaFontStretchIssue](https://github.com/kurobirds/AvaloniaFontStretchIssue)? You can change the `AvaloniaVersion` in `Directory.Build.props` and rebuild for comparison.

**Comment by AngryCarrot789** [comment]
What version will this be officially released in? It doesn't appear to be in 11.1.1 since I still have text rendering issues on fonts like Consolas, but `11.2.999-cibuild0048720-alpha` does work properly

**Comment by Gillibald** [comment]
This will be part of the next release

**Pull request merged**